io.quarkus.kubernetes.deployment.OpenshiftDeployer throws java.io.IOException: Failed to resolve <host> but evidently did reach the host

### Describe the bug

Since recently, with Quarkus 3.0.Alpha4, using `quarkus.kubernetes.deploy=true` leads to an error:

```
Error:  Failed to execute goal io.quarkus:quarkus-maven-plugin:3.0.0.Alpha4:build (default) on project quarkus-github-lottery: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
Error: [ERROR] 	[error]: Build step io.quarkus.kubernetes.deployment.OpenshiftDeployer#checkEnvironment threw an exception: java.lang.RuntimeException: Failed to configure OpenShift. Make sure you have the Quarkus OpenShift extension.
Error:  	at io.quarkus.kubernetes.deployment.OpenshiftDeployer.lambda$checkEnvironment$0(OpenshiftDeployer.java:35)
Error:  	at java.base/java.util.Optional.ifPresent(Optional.java:178)
Error:  	at io.quarkus.kubernetes.deployment.OpenshiftDeployer.checkEnvironment(OpenshiftDeployer.java:22)
Error:  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Error:  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
Error:  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Error:  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
Error:  	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
Error:  	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
Error:  	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
Error:  	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
Error:  	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
Error:  	at java.base/java.lang.Thread.run(Thread.java:833)
Error:  	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Error:  Caused by: io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
Error:  	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:129)
Error:  	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:122)
Error:  	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.restCall(OperationSupport.java:791)
Error:  	at io.fabric8.kubernetes.client.impl.BaseClient.getApiGroups(BaseClient.java:241)
Error:  	at io.fabric8.kubernetes.client.impl.BaseClient.hasApiGroup(BaseClient.java:164)
Error:  	at io.fabric8.openshift.client.impl.OpenShiftClientImpl.isSupported(OpenShiftClientImpl.java:742)
Error:  	at io.quarkus.kubernetes.deployment.OpenshiftDeployer.lambda$checkEnvironment$0(OpenshiftDeployer.java:28)
Error:  	... 13 more
Error:  Caused by: java.io.IOException: Failed to resolve 'api.ospo-osci.z3b1.p1.openshiftapps.com' and search domain query for configured domains failed as well: [lxlrj1mao4menpo2mo3udeoyab.bx.internal.cloudapp.net]
Error:  	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:535)
Error:  	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:558)
Error:  	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.restCall(OperationSupport.java:784)
Error:  	... 17 more
Error:  Caused by: io.netty.resolver.dns.DnsResolveContext$SearchDomainUnknownHostException: Failed to resolve 'api.ospo-osci.z3b1.p1.openshiftapps.com' and search domain query for configured domains failed as well: [lxlrj1mao4menpo2mo3udeoyab.bx.internal.cloudapp.net]
Error:  	at io.netty.resolver.dns.DnsResolveContext.finishResolve(DnsResolveContext.java:1069)
Error:  	at io.netty.resolver.dns.DnsResolveContext.tryToFinishResolve(DnsResolveContext.java:1022)
Error:  	at io.netty.resolver.dns.DnsResolveContext.query(DnsResolveContext.java:418)
Error:  	at io.netty.resolver.dns.DnsResolveContext.access$600(DnsResolveContext.java:66)
Error:  	at io.netty.resolver.dns.DnsResolveContext$2.operationComplete(DnsResolveContext.java:489)
Error:  	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:590)
Error:  	at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:583)
Error:  	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:559)
Error:  	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:492)
Error:  	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:636)
Error:  	at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:629)
Error:  	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:118)
Error:  	at io.netty.resolver.dns.DnsQueryContext.tryFailure(DnsQueryContext.java:256)
Error:  	at io.netty.resolver.dns.DnsQueryContext.onQueryWriteCompletion(DnsQueryContext.java:193)
Error:  	at io.netty.resolver.dns.DnsQueryContext.access$200(DnsQueryContext.java:43)
Error:  	at io.netty.resolver.dns.InflightNameResolver.resolve(InflightNameResolver.java:66)
Error:  	at io.netty.resolver.dns.InflightNameResolver.resolve(InflightNameResolver.java:51)
Error:  	at io.netty.resolver.InetSocketAddressResolver.doResolve(InetSocketAddressResolver.java:55)
Error:  	at io.netty.resolver.InetSocketAddressResolver.doResolve(InetSocketAddressResolver.java:31)
Error:  	at io.netty.resolver.AbstractAddressResolver.resolve(AbstractAddressResolver.java:106)
Error:  	at io.netty.bootstrap.Bootstrap.doResolveAndConnect0(Bootstrap.java:206)
Error:  	at io.netty.bootstrap.Bootstrap.doResolveAndConnect(Bootstrap.java:162)
Error:  	at io.netty.bootstrap.Bootstrap.connect(Bootstrap.java:139)
Error:  	at io.vertx.core.net.impl.ChannelProvider.handleConnect(ChannelProvider.java:155)
Error:  	at io.vertx.core.net.impl.ChannelProvider.connect(ChannelProvider.java:106)
Error:  	at io.vertx.core.net.impl.ChannelProvider.connect(ChannelProvider.java:92)
Error:  	at io.vertx.core.net.impl.NetClientImpl.connectInternal2(NetClientImpl.java:273)
Error:  	at io.vertx.core.net.impl.NetClientImpl.lambda$connectInternal$1(NetClientImpl.java:240)
Error:  	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
Error:  	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
Error:  	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
Error:  	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
Error:  	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
Error:  	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
Error:  	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
Error:  	at io.vertx.core.impl.future.FixedMapping.onSuccess(FixedMapping.java:31)
Error:  	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
Error:  	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
Error:  	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
Error:  	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
Error:  	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
Error:  	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
Error:  	at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62)
Error:  	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
Error:  	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
Error:  	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
Error:  	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
Error:  	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
Error:  	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
Error:  	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
Error:  	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
Error:  	at java.base/java.lang.Thread.run(Thread.java:833): Unknown host Failed to resolve 'api.ospo-osci.z3b1.p1.openshiftapps.com' and search domain query for configured domains failed as well: [lxlrj1mao4menpo2mo3udeoyab.bx.internal.cloudapp.net]
Error:  -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error: Process completed with exit code 1.
```

See https://github.com/quarkusio/quarkus-github-lottery/actions/runs/4290767409/jobs/7479679333#step:6:30555

OpenShift seems to be reachable though, since I can see the deployment causes changes in OpenShift (a build is triggered and the pod restarts). It's stuck on an older version of my application, though.

It's probably a coincidence, but this failure started started to appear right after I upgraded to Maven Compiler Plugin 3.11.0: https://github.com/quarkusio/quarkus-github-lottery/commit/2f75ba23b8d19547eb5ba94fb461adc2e3e699d6


### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

I don't have a reproducer but here is the project: https://github.com/quarkusio/quarkus-github-lottery/

Command:

```
mvn clean package -Dquarkus.kubernetes.deploy=true -Dquarkus.native.container-build=true -Dnative -Drevision=${{ github.sha }}
```

See https://github.com/quarkusio/quarkus-github-lottery/blob/4a0f8d6655025fffa099396854de89d2fc3c1d73/.github/workflows/deploy.yml#L36-L43

Configuration:

```
quarkus.application.name=quarkus-github-lottery
quarkus.application.version=${buildNumber:999-SNAPSHOT}

quarkus.qute.suffixes=md
quarkus.qute.content-types."md"=text/markdown

%test.quarkus.scheduler.enabled=false
%test.quarkus.log.category."io.quarkus.github.lottery".level=DEBUG

%dev.quarkus.scheduler.enabled=false
%dev.quarkus.log.category."io.quarkus.github.lottery".level=DEBUG

%prod.quarkus.openshift.labels."app"=quarkus-github-lottery
# renew the SSL certificate automatically
%prod.quarkus.openshift.annotations."kubernetes.io/tls-acme"=true
%prod.quarkus.openshift.env.configmaps=quarkus-github-lottery-config
%prod.quarkus.openshift.env.secrets=quarkus-github-lottery-secrets
```

See https://github.com/quarkusio/quarkus-github-lottery/blob/4a0f8d6655025fffa099396854de89d2fc3c1d73/src/main/resources/application.properties

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.0.0.Alpha4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/container-image/container-image-openshift/deployment/src/main/java/io/quarkus/container/image/openshift/deployment/OpenshiftProcessor.java
+++ b/extensions/container-image/container-image-openshift/deployment/src/main/java/io/quarkus/container/image/openshift/deployment/OpenshiftProcessor.java
@@ -28,7 +28,6 @@ import java.util.stream.Stream;
 
 import org.jboss.logging.Logger;
 
-import io.dekorate.utils.Clients;
 import io.dekorate.utils.Packaging;
 import io.dekorate.utils.Serialization;
 import io.fabric8.kubernetes.api.model.HasMetadata;
@@ -36,8 +35,10 @@ import io.fabric8.kubernetes.api.model.KubernetesList;
 import io.fabric8.kubernetes.api.model.Secret;
 import io.fabric8.kubernetes.client.Config;
 import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientBuilder;
 import io.fabric8.kubernetes.client.KubernetesClientException;
 import io.fabric8.kubernetes.client.dsl.LogWatch;
+import io.fabric8.kubernetes.client.http.HttpClient;
 import io.fabric8.openshift.api.model.Build;
 import io.fabric8.openshift.api.model.BuildConfig;
 import io.fabric8.openshift.api.model.ImageStream;
@@ -231,7 +232,7 @@ public class OpenshiftProcessor {
     public void openshiftBuildFromJar(OpenshiftConfig openshiftConfig,
             S2iConfig s2iConfig,
             ContainerImageConfig containerImageConfig,
-            KubernetesClientBuildItem kubernetesClientSupplier,
+            KubernetesClientBuildItem kubernetesClientBuilder,
             ContainerImageInfoBuildItem containerImage,
             ArchiveRootBuildItem archiveRoot, OutputTargetBuildItem out, PackageConfig packageConfig,
             List<GeneratedFileSystemResourceBuildItem> generatedResources,
@@ -263,7 +264,7 @@ public class OpenshiftProcessor {
             return;
         }
 
-        try (KubernetesClient kubernetesClient = kubernetesClientSupplier.getClient().get()) {
+        try (KubernetesClient kubernetesClient = kubernetesClientBuilder.buildClient()) {
             String namespace = Optional.ofNullable(kubernetesClient.getNamespace()).orElse("default");
             LOG.info("Starting (in-cluster) container image build for jar using: " + config.buildStrategy + " on server: "
                     + kubernetesClient.getMasterUrl() + " in namespace:" + namespace + ".");
@@ -272,14 +273,16 @@ public class OpenshiftProcessor {
             //For s2i kind of builds where jars are expected directly in the '/' we have to use null.
             String outputDirName = out.getOutputDirectory().getFileName().toString();
             String contextRoot = getContextRoot(outputDirName, packageConfig.isFastJar(), config.buildStrategy);
+            KubernetesClientBuilder clientBuilder = newClientBuilderWithoutHttp2(kubernetesClient.getConfiguration(),
+                    kubernetesClientBuilder.getHttpClientFactory());
             if (packageConfig.isFastJar()) {
-                createContainerImage(kubernetesClient, openshiftYml.get(), config, contextRoot, jar.getPath().getParent(),
+                createContainerImage(clientBuilder, openshiftYml.get(), config, contextRoot, jar.getPath().getParent(),
                         jar.getPath().getParent());
             } else if (jar.getLibraryDir() != null) { //When using uber-jar the libraryDir is going to be null, potentially causing NPE.
-                createContainerImage(kubernetesClient, openshiftYml.get(), config, contextRoot, jar.getPath().getParent(),
+                createContainerImage(clientBuilder, openshiftYml.get(), config, contextRoot, jar.getPath().getParent(),
                         jar.getPath(), jar.getLibraryDir());
             } else {
-                createContainerImage(kubernetesClient, openshiftYml.get(), config, contextRoot, jar.getPath().getParent(),
+                createContainerImage(clientBuilder, openshiftYml.get(), config, contextRoot, jar.getPath().getParent(),
                         jar.getPath());
             }
             artifactResultProducer.produce(new ArtifactResultBuildItem(null, "jar-container", Collections.emptyMap()));
@@ -300,7 +303,7 @@ public class OpenshiftProcessor {
     @BuildStep(onlyIf = { IsNormalNotRemoteDev.class, OpenshiftBuild.class, NativeBuild.class })
     public void openshiftBuildFromNative(OpenshiftConfig openshiftConfig, S2iConfig s2iConfig,
             ContainerImageConfig containerImageConfig,
-            KubernetesClientBuildItem kubernetesClientSupplier,
+            KubernetesClientBuildItem kubernetesClientBuilder,
             ContainerImageInfoBuildItem containerImage,
             ArchiveRootBuildItem archiveRoot, OutputTargetBuildItem out, PackageConfig packageConfig,
             List<GeneratedFileSystemResourceBuildItem> generatedResources,
@@ -321,7 +324,7 @@ public class OpenshiftProcessor {
             return;
         }
 
-        try (KubernetesClient kubernetesClient = kubernetesClientSupplier.getClient().get()) {
+        try (KubernetesClient kubernetesClient = kubernetesClientBuilder.buildClient()) {
             String namespace = Optional.ofNullable(kubernetesClient.getNamespace()).orElse("default");
 
             LOG.info("Starting (in-cluster) container image build for jar using: " + config.buildStrategy + " on server: "
@@ -340,14 +343,16 @@ public class OpenshiftProcessor {
             //For docker kind of builds where we use instructions like: `COPY target/*.jar /deployments` it using '/target' is a requirement.
             //For s2i kind of builds where jars are expected directly in the '/' we have to use null.
             String contextRoot = config.buildStrategy == BuildStrategy.DOCKER ? "target" : null;
-            createContainerImage(kubernetesClient, openshiftYml.get(), config, contextRoot, out.getOutputDirectory(),
-                    nativeImage.getPath());
+            createContainerImage(
+                    newClientBuilderWithoutHttp2(kubernetesClient.getConfiguration(),
+                            kubernetesClientBuilder.getHttpClientFactory()),
+                    openshiftYml.get(), config, contextRoot, out.getOutputDirectory(), nativeImage.getPath());
             artifactResultProducer.produce(new ArtifactResultBuildItem(null, "native-container", Collections.emptyMap()));
             containerImageBuilder.produce(new ContainerImageBuilderBuildItem(OPENSHIFT));
         }
     }
 
-    public static void createContainerImage(KubernetesClient kubernetesClient,
+    public static void createContainerImage(KubernetesClientBuilder kubernetesClientBuilder,
             GeneratedFileSystemResourceBuildItem openshiftManifests,
             OpenshiftConfig openshiftConfig,
             String base,
@@ -364,10 +369,7 @@ public class OpenshiftProcessor {
             throw new RuntimeException("Error creating the openshift binary build archive.", e);
         }
 
-        Config config = kubernetesClient.getConfiguration();
-        //Let's disable http2 as it causes issues with duplicate build triggers.
-        config.setHttp2Disable(true);
-        try (KubernetesClient client = Clients.fromConfig(config)) {
+        try (KubernetesClient client = kubernetesClientBuilder.build()) {
             OpenShiftClient openShiftClient = toOpenshiftClient(client);
             KubernetesList kubernetesList = Serialization
                     .unmarshalAsList(new ByteArrayInputStream(openshiftManifests.getData()));
@@ -531,6 +533,14 @@ public class OpenshiftProcessor {
         for (String line = reader.readLine(); line != null; line = reader.readLine()) {
             LOG.log(level, line);
         }
+    }
+
+    private static KubernetesClientBuilder newClientBuilderWithoutHttp2(Config configuration,
+            HttpClient.Factory httpClientFactory) {
+        //Let's disable http2 as it causes issues with duplicate build triggers.
+        configuration.setHttp2Disable(true);
+
+        return new KubernetesClientBuilder().withConfig(configuration).withHttpClientFactory(httpClientFactory);
     }
 
     // visible for test

--- a/extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/S2iProcessor.java
+++ b/extensions/container-image/container-image-s2i/deployment/src/main/java/io/quarkus/container/image/s2i/deployment/S2iProcessor.java
@@ -171,7 +171,7 @@ public class S2iProcessor {
 
     @BuildStep(onlyIf = { IsNormalNotRemoteDev.class, S2iBuild.class }, onlyIfNot = NativeBuild.class)
     public void s2iBuildFromJar(S2iConfig s2iConfig, ContainerImageConfig containerImageConfig,
-            KubernetesClientBuildItem kubernetesClientSupplier,
+            KubernetesClientBuildItem kubernetesClientBuilder,
             ContainerImageInfoBuildItem containerImage,
             ArchiveRootBuildItem archiveRoot, OutputTargetBuildItem out, PackageConfig packageConfig,
             List<GeneratedFileSystemResourceBuildItem> generatedResources,
@@ -201,7 +201,7 @@ public class S2iProcessor {
                     "No Openshift manifests were generated so no s2i process will be taking place");
             return;
         }
-        try (KubernetesClient kubernetesClient = kubernetesClientSupplier.getClient().get()) {
+        try (KubernetesClient kubernetesClient = kubernetesClientBuilder.buildClient()) {
             String namespace = Optional.ofNullable(kubernetesClient.getNamespace()).orElse("default");
             LOG.info("Performing s2i binary build with jar on server: " + kubernetesClient.getMasterUrl()
                     + " in namespace:" + namespace + ".");
@@ -215,7 +215,7 @@ public class S2iProcessor {
 
     @BuildStep(onlyIf = { IsNormalNotRemoteDev.class, S2iBuild.class, NativeBuild.class })
     public void s2iBuildFromNative(S2iConfig s2iConfig, ContainerImageConfig containerImageConfig,
-            KubernetesClientBuildItem kubernetesClientSupplier,
+            KubernetesClientBuildItem kubernetesClientBuilder,
             ContainerImageInfoBuildItem containerImage,
             ArchiveRootBuildItem archiveRoot, OutputTargetBuildItem out, PackageConfig packageConfig,
             List<GeneratedFileSystemResourceBuildItem> generatedResources,
@@ -234,7 +234,7 @@ public class S2iProcessor {
             return;
         }
 
-        try (KubernetesClient kubernetesClient = kubernetesClientSupplier.getClient().get()) {
+        try (KubernetesClient kubernetesClient = kubernetesClientBuilder.buildClient()) {
             String namespace = Optional.ofNullable(kubernetesClient.getNamespace()).orElse("default");
             LOG.info("Performing s2i binary build with native image on server: " + kubernetesClient.getMasterUrl()
                     + " in namespace:" + namespace + ".");

--- a/extensions/kubernetes-client/deployment-internal/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientBuildStep.java
+++ b/extensions/kubernetes-client/deployment-internal/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientBuildStep.java
@@ -1,9 +1,11 @@
 package io.quarkus.kubernetes.client.deployment;
 
-import static io.quarkus.kubernetes.client.runtime.KubernetesClientUtils.*;
+import static io.quarkus.kubernetes.client.runtime.KubernetesClientUtils.createConfig;
 
 import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.QuarkusBuildCloseablesBuildItem;
 import io.quarkus.kubernetes.client.runtime.KubernetesClientBuildConfig;
+import io.quarkus.kubernetes.client.runtime.QuarkusHttpClientFactory;
 import io.quarkus.kubernetes.client.spi.KubernetesClientBuildItem;
 import io.quarkus.runtime.TlsConfig;
 
@@ -12,7 +14,9 @@ public class KubernetesClientBuildStep {
     private KubernetesClientBuildConfig buildConfig;
 
     @BuildStep
-    public KubernetesClientBuildItem process(TlsConfig tlsConfig) {
-        return new KubernetesClientBuildItem(createConfig(buildConfig, tlsConfig));
+    public KubernetesClientBuildItem process(TlsConfig tlsConfig, QuarkusBuildCloseablesBuildItem closeablesBuildItem) {
+        QuarkusHttpClientFactory httpClientFactory = new QuarkusHttpClientFactory();
+        closeablesBuildItem.add(httpClientFactory);
+        return new KubernetesClientBuildItem(createConfig(buildConfig, tlsConfig), httpClientFactory);
     }
 }

--- a/extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/QuarkusHttpClientFactory.java
+++ b/extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/QuarkusHttpClientFactory.java
@@ -2,6 +2,8 @@ package io.quarkus.kubernetes.client.runtime;
 
 import static io.vertx.core.spi.resolver.ResolverProvider.DISABLE_DNS_RESOLVER_PROP_NAME;
 
+import java.io.Closeable;
+
 import jakarta.enterprise.inject.spi.CDI;
 
 import io.fabric8.kubernetes.client.Config;
@@ -10,17 +12,20 @@ import io.fabric8.kubernetes.client.vertx.VertxHttpClientBuilder;
 import io.vertx.core.Vertx;
 import io.vertx.core.VertxOptions;
 
-public class QuarkusHttpClientFactory implements HttpClient.Factory {
+public class QuarkusHttpClientFactory implements HttpClient.Factory, Closeable {
 
     private Vertx vertx;
+    private boolean closeVertxOnExit;
 
     public QuarkusHttpClientFactory() {
         // The client might get initialized outside a Quarkus context that can provide the Vert.x instance
         // This is the case for the MockServer / @KubernetesTestServer where the server provides the KubernetesClient instance
         try {
             this.vertx = CDI.current().select(Vertx.class).get();
+            this.closeVertxOnExit = false;
         } catch (Exception e) {
             this.vertx = createVertxInstance();
+            this.closeVertxOnExit = true;
         }
     }
 
@@ -58,5 +63,12 @@ public class QuarkusHttpClientFactory implements HttpClient.Factory {
     @Override
     public int priority() {
         return 1;
+    }
+
+    @Override
+    public void close() {
+        if (closeVertxOnExit) {
+            vertx.close();
+        }
     }
 }

--- a/extensions/kubernetes-client/runtime-internal/src/main/resources/META-INF/services/io.fabric8.kubernetes.client.http.HttpClient$Factory
+++ b/extensions/kubernetes-client/runtime-internal/src/main/resources/META-INF/services/io.fabric8.kubernetes.client.http.HttpClient$Factory
@@ -1,1 +1,0 @@
-io.quarkus.kubernetes.client.runtime.QuarkusHttpClientFactory

--- a/extensions/kubernetes-client/spi/src/main/java/io/quarkus/kubernetes/client/spi/KubernetesClientBuildItem.java
+++ b/extensions/kubernetes-client/spi/src/main/java/io/quarkus/kubernetes/client/spi/KubernetesClientBuildItem.java
@@ -1,26 +1,30 @@
 package io.quarkus.kubernetes.client.spi;
 
-import java.util.function.Supplier;
-
 import io.fabric8.kubernetes.client.Config;
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.fabric8.kubernetes.client.KubernetesClientBuilder;
+import io.fabric8.kubernetes.client.http.HttpClient;
 import io.quarkus.builder.item.SimpleBuildItem;
 
 public final class KubernetesClientBuildItem extends SimpleBuildItem {
 
     private final Config config;
+    private final HttpClient.Factory httpClientFactory;
 
-    public KubernetesClientBuildItem(Config config) {
+    public KubernetesClientBuildItem(Config config, HttpClient.Factory httpClientFactory) {
         this.config = config;
-    }
-
-    public Supplier<KubernetesClient> getClient() {
-        return () -> new KubernetesClientBuilder().withConfig(config).build();
+        this.httpClientFactory = httpClientFactory;
     }
 
     public Config getConfig() {
         return config;
     }
 
+    public HttpClient.Factory getHttpClientFactory() {
+        return httpClientFactory;
+    }
+
+    public KubernetesClient buildClient() {
+        return new KubernetesClientBuilder().withConfig(config).withHttpClientFactory(httpClientFactory).build();
+    }
 }

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java
@@ -17,13 +17,14 @@ public class KnativeDeployer {
     @BuildStep
     public void checkEnvironment(Optional<SelectedKubernetesDeploymentTargetBuildItem> selectedDeploymentTarget,
             List<GeneratedKubernetesResourceBuildItem> resources,
-            KubernetesClientBuildItem clientSupplier, BuildProducer<KubernetesDeploymentClusterBuildItem> deploymentCluster) {
+            KubernetesClientBuildItem kubernetesClientBuilder,
+            BuildProducer<KubernetesDeploymentClusterBuildItem> deploymentCluster) {
         selectedDeploymentTarget.ifPresent(target -> {
-            if (!KubernetesDeploy.INSTANCE.checkSilently()) {
+            if (!KubernetesDeploy.INSTANCE.checkSilently(kubernetesClientBuilder)) {
                 return;
             }
             if (target.getEntry().getName().equals(KNATIVE)) {
-                try (DefaultKnativeClient client = clientSupplier.getClient().get().adapt(DefaultKnativeClient.class)) {
+                try (DefaultKnativeClient client = kubernetesClientBuilder.buildClient().adapt(DefaultKnativeClient.class)) {
                     if (client.isSupported()) {
                         deploymentCluster.produce(new KubernetesDeploymentClusterBuildItem(KNATIVE));
                     } else {

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeploy.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeploy.java
@@ -9,7 +9,7 @@ import org.jboss.logging.Logger;
 
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.fabric8.kubernetes.client.VersionInfo;
-import io.quarkus.kubernetes.client.runtime.KubernetesClientUtils;
+import io.quarkus.kubernetes.client.spi.KubernetesClientBuildItem;
 
 public class KubernetesDeploy {
 
@@ -27,8 +27,8 @@ public class KubernetesDeploy {
      *
      * @throws RuntimeException if there was an error while communicating with the Kubernetes API server
      */
-    public boolean check() {
-        Result result = doCheck();
+    public boolean check(KubernetesClientBuildItem clientBuilder) {
+        Result result = doCheck(clientBuilder);
 
         if (result.getException().isPresent()) {
             throw result.getException().get();
@@ -41,11 +41,11 @@ public class KubernetesDeploy {
      * @return {@code true} if @{code quarkus.kubernetes.deploy=true} AND the target Kubernetes API server is reachable
      *         {@code false} otherwise or if there was an error while communicating with the Kubernetes API server
      */
-    public boolean checkSilently() {
-        return doCheck().isAllowed();
+    public boolean checkSilently(KubernetesClientBuildItem clientBuilder) {
+        return doCheck(clientBuilder).isAllowed();
     }
 
-    private Result doCheck() {
+    private Result doCheck(KubernetesClientBuildItem clientBuilder) {
         if (!KubernetesConfigUtil.isDeploymentEnabled()) {
             return Result.notConfigured();
         }
@@ -55,9 +55,8 @@ public class KubernetesDeploy {
             return Result.enabled();
         }
 
-        KubernetesClient client = KubernetesClientUtils.createClient();
-        String masterURL = client.getConfiguration().getMasterUrl();
-        try {
+        String masterURL = clientBuilder.getConfig().getMasterUrl();
+        try (KubernetesClient client = clientBuilder.buildClient()) {
             //Let's check if we can connect.
             VersionInfo version = client.getVersion();
             if (version == null) {
@@ -83,8 +82,6 @@ public class KubernetesDeploy {
                                 + masterURL + "'",
                         e));
             }
-        } finally {
-            client.close();
         }
     }
 

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployer.java
@@ -23,7 +23,6 @@ import java.util.stream.Collectors;
 
 import org.jboss.logging.Logger;
 
-import io.dekorate.utils.Clients;
 import io.dekorate.utils.Serialization;
 import io.fabric8.kubernetes.api.model.APIResourceList;
 import io.fabric8.kubernetes.api.model.GenericKubernetesResource;
@@ -90,11 +89,11 @@ public class KubernetesDeployer {
 
     @BuildStep
     public void checkEnvironment(Optional<SelectedKubernetesDeploymentTargetBuildItem> selectedDeploymentTarget,
-            KubernetesClientBuildItem client,
+            KubernetesClientBuildItem kubernetesClientBuilder,
             List<GeneratedKubernetesResourceBuildItem> resources,
             BuildProducer<KubernetesDeploymentClusterBuildItem> deploymentCluster) {
 
-        if (!KubernetesDeploy.INSTANCE.checkSilently()) {
+        if (!KubernetesDeploy.INSTANCE.checkSilently(kubernetesClientBuilder)) {
             return;
         }
         String target = selectedDeploymentTarget.map(s -> s.getEntry().getName()).orElse(KUBERNETES);
@@ -104,7 +103,7 @@ public class KubernetesDeployer {
     }
 
     @BuildStep(onlyIf = IsNormalNotRemoteDev.class)
-    public void deploy(KubernetesClientBuildItem kubernetesClientSupplier,
+    public void deploy(KubernetesClientBuildItem kubernetesClientBuilder,
             Capabilities capabilities,
             List<KubernetesDeploymentClusterBuildItem> deploymentClusters,
             Optional<SelectedKubernetesDeploymentTargetBuildItem> selectedDeploymentTarget,
@@ -117,7 +116,7 @@ public class KubernetesDeployer {
             // needed to ensure that this step runs after the container image has been built
             @SuppressWarnings("unused") List<ArtifactResultBuildItem> artifactResults) {
 
-        if (!KubernetesDeploy.INSTANCE.check()) {
+        if (!KubernetesDeploy.INSTANCE.check(kubernetesClientBuilder)) {
             return;
         }
 
@@ -134,7 +133,7 @@ public class KubernetesDeployer {
             return;
         }
 
-        try (final KubernetesClient client = Clients.fromConfig(kubernetesClientSupplier.getConfig())) {
+        try (final KubernetesClient client = kubernetesClientBuilder.buildClient()) {
             deploymentResult
                     .produce(deploy(selectedDeploymentTarget.get().getEntry(), client, outputTarget.getOutputDirectory(),
                             openshiftConfig, applicationInfo, optionalResourceDefinitions));

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployerPrerequisite.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesDeployerPrerequisite.java
@@ -10,11 +10,13 @@ import io.quarkus.container.spi.FallbackContainerImageRegistryBuildItem;
 import io.quarkus.deployment.IsNormalNotRemoteDev;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.kubernetes.client.spi.KubernetesClientBuildItem;
 
 public class KubernetesDeployerPrerequisite {
 
     @BuildStep(onlyIf = IsNormalNotRemoteDev.class)
     public void prepare(ContainerImageInfoBuildItem containerImage,
+            KubernetesClientBuildItem kubernetesClientBuilder,
             Optional<SelectedKubernetesDeploymentTargetBuildItem> selectedDeploymentTarget,
             Optional<FallbackContainerImageRegistryBuildItem> fallbackRegistry,
             List<PreventImplicitContainerImagePushBuildItem> preventImplicitContainerImagePush,
@@ -23,7 +25,7 @@ public class KubernetesDeployerPrerequisite {
 
         // we don't want to throw an exception at this step and fail the build because it could prevent
         // the Kubernetes resources from being generated
-        if (!KubernetesDeploy.INSTANCE.checkSilently() || !selectedDeploymentTarget.isPresent()) {
+        if (!KubernetesDeploy.INSTANCE.checkSilently(kubernetesClientBuilder) || !selectedDeploymentTarget.isPresent()) {
             return;
         }
 

--- a/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftDeployer.java
+++ b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftDeployer.java
@@ -18,13 +18,14 @@ public class OpenshiftDeployer {
     @BuildStep
     public void checkEnvironment(Optional<SelectedKubernetesDeploymentTargetBuildItem> selectedDeploymentTarget,
             List<GeneratedKubernetesResourceBuildItem> resources,
-            KubernetesClientBuildItem clientSupplier, BuildProducer<KubernetesDeploymentClusterBuildItem> deploymentCluster) {
+            KubernetesClientBuildItem kubernetesClientBuilder,
+            BuildProducer<KubernetesDeploymentClusterBuildItem> deploymentCluster) {
         selectedDeploymentTarget.ifPresent(target -> {
-            if (!KubernetesDeploy.INSTANCE.checkSilently()) {
+            if (!KubernetesDeploy.INSTANCE.checkSilently(kubernetesClientBuilder)) {
                 return;
             }
             if (target.getEntry().getName().equals(OPENSHIFT)) {
-                try (var openShiftClient = clientSupplier.getClient().get().adapt(OpenShiftClient.class)) {
+                try (var openShiftClient = kubernetesClientBuilder.buildClient().adapt(OpenShiftClient.class)) {
                     if (openShiftClient.isSupported()) {
                         deploymentCluster.produce(new KubernetesDeploymentClusterBuildItem(OPENSHIFT));
                     } else {

